### PR TITLE
[ENG-5997] [2] update vgs related code to comply with security measures

### PR DESCRIPTION
**Pull request by tonatiuh**
What this PR does: 
* Updates code so that client_id and client_secret are not sent to VGS proxy by any means 
* Avoids logging the VGS proxy URL
* Introduces the Typhoeus gem to be able to generate an HTTPS connection to proxies (to prevent the username/password in the proxy URL from traveling unencrypted)
* Completes the functionality to successfully connect to proxies (with support for custom certificate authority pem and HTTPS proxy URL)

### CODE REVIEW

File: lib/synapse_pay_rest/http_client.rb [hunk 1]
```diff
@@ -19,15 +20,14 @@ class HTTPClient
...
       log_to         = options[:log_to] || 'stdout'
       RestClient.log = log_to if options[:logging]
       @logging       = options[:logging]
       @development_mode = development_mode
 
-      @vgs_url = vgs_url
-      @vgs_sandbox_pem = options[:sandbox_pem]
-      @vgs_live_pem = options[:live_pem]
+      @vgs_url = options[:vgs_url]
+      @vgs_pem = options[:vgs_pem]
```
Comment (tonatiuh) [inline lib/synapse_pay_rest/http_client.rb]: With this change, the logic to decide what vgs_url and vgs_pem to use per env must be handled by the app using this gem, that way this gem just uses whatever values are passed.

File: lib/synapse_pay_rest/http_client.rb [hunk 2]
```diff
@@ -85,44 +78,35 @@ def update_headers(oauth_key: nil, fingerprint: nil, client_id: nil,
...
 
-      params = {
-        method: method,
-        url: url,
-        headers: request_headers
-      }
+    def vgs_proxy_request_params(base_params:)
+      base_params.merge(
+        ssl_client_cert: OpenSSL::X509::Certificate.new(vgs_pem),
+        verify_ssl: OpenSSL::SSL::VERIFY_NONE,
```
Comment (lukeasrodgers) [inline lib/synapse_pay_rest/http_client.rb]: it looks like this is copied from old code, but there is generally no good reason to do this in production that I'm aware of. maybe in dev/test if using self-signed certs okay, but i think we want `OpenSSL::SSL::VERIFY_PEER` instead here (or probably just leaving it blank will get that by default, but doesn't hurt to be explicit here)

File: lib/synapse_pay_rest/http_client.rb [hunk 3]
```diff
@@ -85,44 +78,35 @@ def update_headers(oauth_key: nil, fingerprint: nil, client_id: nil,
...
+        proxy: vgs_url,
 
-      params[:payload] = payload.to_json if payload.present?
-      params.merge!(tunnel_params) if tunnel
-      params
+        # NOTE: next line is just defensive programming, making sure the client_id and client_secret are not sent to VGS.
```
Comment (lukeasrodgers) [inline lib/synapse_pay_rest/http_client.rb]: this is great, but can i suggest a slight language change: remove the "just" and make the wording more forceful, like "under no circumstances do we want to send the client_id and client_secret to VGS"

**Review by lukeasrodgers — CHANGES_REQUESTED** [review]
couple mostly minor things, also could we have at least one sanity-check test for this? i would like to see something like "passing in `tunnel: true` to a POST request results in a request via the VGS proxy". you could use webmock like I did here https://github.com/alice-financial/SynapsePayRest-Ruby/pull/17/files for example

File: lib/synapse_pay_rest/client.rb [hunk 4]
```diff
@@ -32,15 +32,12 @@ class Client
...
       base_url = development_mode ? 'https://uat-api.synapsefi.com/v3.1' : 'https://api.synapsefi.com/v3.1'
-      vgs_url = development_mode ? options[:sandbox_url] : options[:live_url]
 
       @http_client  = HTTPClient.new(base_url: base_url,
                                      client_id: client_id,
                                      client_secret: client_secret,
                                      fingerprint: fingerprint,
                                      ip_address: ip_address,
-                                     development_mode: development_mode,
-                                     vgs_url: vgs_url,
```
Comment (tonatiuh) [inline lib/synapse_pay_rest/client.rb:43]: The new param comes as `vgs_proxy_config` in the `options` hash.

File: lib/synapse_pay_rest/http_client.rb [hunk 5]
```diff
@@ -85,54 +81,22 @@ def update_headers(oauth_key: nil, fingerprint: nil, client_id: nil,
...
     # @param path [String]
     # @param payload [Hash]
     # @param idempotency_key [String] (optional) avoid accidentally performing the same operation twice
     #
     # @raise [SynapsePayRest::Error] subclass depends on HTTP response
-    # 
+    #
     # @return [Hash] API response
     def post(path, payload, **options)
-      tunnel = options.delete(:tunnel) || false
```
Comment (tonatiuh) [inline lib/synapse_pay_rest/http_client.rb:128]: We're not proxying any request for post method. So removing this line.

File: lib/synapse_pay_rest/models/subnet/subnet.rb [hunk 6]
```diff
@@ -41,14 +41,14 @@ def create(node:, nickname:, **options)
...
       # @param node [SynapsePayRest::BaseNode] node to which the subnet belongs
       # @param id [String] id of the subnet to find
-      # 
+      #
       # @raise [SynapsePayRest::Error] if not found or other HTTP error
-      # 
+      #
       # @return [SynapsePayRest::Subnet]
-      def find(node:, id:, full_dehydrate: false)
+      def find(node:, id:, full_dehydrate: "no")
```
Comment (tonatiuh) [inline lib/synapse_pay_rest/models/subnet/subnet.rb:51]: Synapse expects a string ("yes"/"no") for the full_dehydrate, a boolean is an invalid param.

**Review by tonatiuh — COMMENTED** [review]
I made changes based on your comments @lukeasrodgers. Also, I added the [typhoeus](https://github.com/typhoeus/typhoeus) gem to support HTTPS connections to proxies (that's to avoid the username/password in the connection URL traveling unencrypted in the requests). 

Can you take another look?

File: lib/synapse_pay_rest/http_client.rb [hunk 7]
```diff
@@ -194,6 +159,69 @@ def delete(path)
...
+      request_headers = headers
+
+      if idempotency_key
+        request_headers.merge!({'X-SP-IDEMPOTENCY-KEY' => idempotency_key})
+      end
+
+      options = {
+        proxy: vgs_proxy_config[:host],
+        proxyuserpwd: "#{vgs_proxy_config[:username]}:#{vgs_proxy_config[:password]}",
+        ssl_verifyhost: 5,
```
Comment (lukeasrodgers) [inline lib/synapse_pay_rest/http_client.rb]: where does this config come from? i'm guessing https://curl.se/libcurl/c/CURLOPT_SSL_VERIFYHOST.html but there is no mention of `5` as a possible value there

File: lib/synapse_pay_rest/http_client.rb [hunk 8]
```diff
@@ -194,6 +159,69 @@ def delete(path)
...
+    end
+
+    # Sends a requests
+    #
+    # @param method [Symbol]
+    # @param url [String]
+    # @param payload [Hash]
+    # @param idempotency_key [String]
+    #
+    # @return [String] Stringified version of the JSON response
```
Comment (lukeasrodgers) [inline lib/synapse_pay_rest/http_client.rb]: are you sure this is the return type? i thought this would return `RestClient::Response`

**Review by lukeasrodgers — COMMENTED** [review]
@tonatiuh just a couple Qs, looks good otherwise

File: lib/synapse_pay_rest/http_client.rb (hunk 7 above)
Comment (tonatiuh) [reply lib/synapse_pay_rest/http_client.rb]: ups, that was a typo, thanks
I've set the right value `2` to perform host verification (which Typhoeus also documents in [ref](https://github.com/typhoeus/typhoeus/blob/d12276f2c7b228fdf090010b70b9b28463ab72af/README.md#ssl)).

**Review by lukeasrodgers — APPROVED** [review]

**Pull request merged**